composer/metrics/nlp.py

```diff
@@ -479,3 +493,117 @@ def update(self, batch: Dict[str, Any], output_logits: torch.Tensor, labels: tor
+
+    In each case, the model constructs a given number of continuations, and each continuation is run against a set of test cases. The model is considered
+    correct if at least one of the proposed continuations passes all the test cases.
+
+    Adds metric state variables:
+        correct (float): The number of instances where the predictions passed all the test cases.
+        total (float): The number of total instances that were predicted.
+
+    Args:
+        dist_sync_on_step (bool, optional): Synchronize metric state across processes at
+            each forward() before returning the value at the step. Default: ``False``.
+    """
+
+    # Make torchmetrics call update only once
+    full_state_update = False
+
+    def __init__(self, dist_sync_on_step: bool = False):
+        # state from multiple processes
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+        self.add_state('correct', default=torch.tensor(0.), dist_reduce_fx='sum')
+        self.add_state('total', default=torch.tensor(0.), dist_reduce_fx='sum')
+
+    def update(self, batch: Dict[str, Any], outputs: List[str], labels: List[str]):
+        del labels  # never used
+        remote = 'LAMBDA_FUNCTION_ARN' in os.environ and 'AWS_DEFAULT_REGION' in os.environ
+        client = boto3.Session().client('lambda') if remote else None
+        if not remote:
+            warnings.warn("Running code eval locally may be unsecure.")
+        num_beams = batch['generation_kwargs']['num_beams']
+        processed_outputs = [outputs[i * num_beams:(i + 1) * num_beams] for i in range(len(batch['prompts']))]
+        for sample_outputs, sample_prompt, test_inputs, test_outputs, entry_point in zip(
+                processed_outputs, batch['prompts'], batch['test_inputs'], batch['test_outputs'],
+                batch['entry_points']):
+            self.total += torch.tensor(1.0)
+            for code_gen in sample_outputs:
+                code_gen = re.split(r'\ndef|\nclass|\n#|\nif|\nprint', code_gen)[0]
+                final_code = sample_prompt + code_gen
+                passes_all = True
+                for test_input, test_output in zip(test_inputs, test_outputs):
+                    if remote:
+                        if not self.update_online_helper(final_code, test_input, test_output, entry_point, client):
+                            passes_all = False
+                            break
+                    else:
+                        ret = multiprocessing.Value('b', 0)
+                        p = multiprocessing.Process(target=self.update_offline_helper,
+                                                    args=(final_code, test_input, test_output, entry_point, ret))
+                        p.start()
+                        p.join(TIMEOUT)
+                        p.terminate()
+                        if not bool(ret.value):
+                            passes_all = False
```

Comment:
Can we put some comments here on the two paths?

Reply:
done